apps/keyboard/js/settings/tst_converter.js

```diff
@@ -0,0 +1,413 @@
+    return this.rotateRight(root);
+  } else if (node.ch > root.ch) {
+    root.right = this.promoteNodeToRoot(root.right, node);
+    return this.rotateLeft(root);
+  } else {
+    return root;
+  }
+};
+
+// balance the whole TST
+TSTTree.prototype.balanceTree = function(node) {
+  if (!node) {
+    return;
+  }
+
+  // promote to root the letter with the highest maximum frequency
+  // of a suffix starting with this letter
+  node = this.promoteNodeToRoot(node, this.sortAndAnnotate(node));
+
+  // balance other letters on this level of the tree
+  node.left = this.balanceLevel(node.left);
+  node.right = this.balanceLevel(node.right);
+  node.center = this.balanceTree(node.center);
+
+  return node;
+};
+
+// balance the whole TST
+TSTTree.prototype.balance = function(root) {
+  this.setCount(root);
+
+  root = this.balanceTree(root);
+
+  return root;
+};
+
+var TSTConverter = function(words) {
+  this.blob = undefined;
+  this.words = words;
+
+  // How many times do we use each character in this language
+  this.characterFrequency = {};
+  this.maxWordLength = 0;
+};
+
+TSTConverter.prototype._DEBUG = false;
+
+TSTConverter.prototype.debug = function(msg) {
+  if (this._DEBUG) {
+    console.log(msg);
+  }
+};
+
+// Serialize the tree to an array. Do it depth first, folling the
+// center pointer first because that might give us better locality
+TSTConverter.prototype.serializeNode = function(node, output) {
+  output.push(node);
+  node.offset = output.length;
+
+  if (node.ch == _EndOfWord && node.center) {
+    this.debug('nul node with a center!');
+  }
+  if (node.ch != _EndOfWord && !node.center) {
+    this.debug('char node with no center!');
+  }
+
+  // do the center node first so words are close together
+  if (node.center) {
+    this.serializeNode(node.center, output);
+  }
+
+  if (node.left) {
+    this.serializeNode(node.left, output);
+  }
+
+  if (node.right) {
+    this.serializeNode(node.right, output);
+  }
+};
+
+TSTConverter.prototype.serializeTree = function(root) {
+  var output = [];
+  this.serializeNode(root, output);
+  return output;
+};
+
+// Make a pass through the array of nodes and figure out the size and offset
+// of each one.
+TSTConverter.prototype.computeOffsets = function(nodes) {
+  var offset = 0;
+
+  nodes.forEach(function(node) {
+    node.offset = offset;
+
+    var charlen;
+    if (node.ch == _EndOfWord) {
+      charlen = 0;
+    } else if (node.ch.charCodeAt(0) <= 255) {
+      charlen = 1;
+    } else {
+      charlen = 2;
+    }
+
+    var nextlen = node.next ? 3 : 0;
+
+    offset = offset + 1 + charlen + nextlen;
+  });
+
+  return offset;
+};
+
+// JSConv:
+// In the JS version, since we're not directly writing to a file,
+// 'output' is a JS array. We convert to UInt8Array when we
+// finishes pushing to 'output'. This is because UInt8Array's length
+// has to be decided at instantiation.
+// XXX: See if we can pre-determine the length when instantiating
+//      the buffer.
```

Comment:
After I've now studied the blob format and generation code more thoroughly, I believe this pre-determination is quite possible. I'll do that in a follow-up bug.